target/riscv/helper.c

```diff
@@ -196,13 +198,26 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             target_ulong updated_pte = pte | PTE_A |
                 (access_type == MMU_DATA_STORE ? PTE_D : 0);
             if (updated_pte != pte) {
-                pte = updated_pte;
-                /* NOTE: this should be atomic e.g. use cmpxchg */
-#if defined (TARGET_RISCV32)
-                stl_phys(cs->as, pte_addr, pte);
-#elif defined (TARGET_RISCV64)
-                stq_phys(cs->as, pte_addr, pte);
-#endif
+                /* if accessed or dirty bits need updating, and the PTE is
+                 * in RAM, then we do so atomically with a compare and swap.
+                 * if the PTE is in IO space, then it can't be updated.
+                 * if the PTE changed, then we must re-walk the page table
+                   as the PTE is no longer valid */
+                MemoryRegion *mr;
+                hwaddr l = sizeof(target_ulong), addr1;
+                mr = address_space_translate(cs->as, pte_addr,
+                    &addr1, &l, false);
+                if (memory_access_is_direct(mr, true)) {
+                    target_ulong *pte_pa =
+                        qemu_map_ram_ptr(mr->ram_block, addr1);
+                    target_ulong old_pte =
+                        atomic_cmpxchg(pte_pa, pte, updated_pte);
+                    if (old_pte != pte) {
+                        goto restart;
+                    } else {
+                        pte = updated_pte;
+                    }
+                }
```

Comment:
There should be an else here which fails the address translation if the PTE cannot be atomically updated.

Reply:
Sure. We can add a TRANSLATE_FAIL here.

Reply:
With my first two comments addressed I'd be fine with this.  A very slight concern is that none of the functions you're calling seem to handle TB invalidation or hardware watchpoints, although having either overlap a page table would be quite strange.